File: 008/cpp/main.cpp
class Solution {
public:
    bool preCheck1(int r, int r0) {
        return r > 214748364 || (r == 214748364 && r0 > 7);
    }

    bool preCheck2(int r, int r0) {
        return r > 214748364 || (r == 214748364 && r0 >= 8);
    }

    bool preCheck(int r, int r0, int sign) {
        return sign > 0 ? preCheck1(r, r0) : preCheck2(r, r0);
    }

    int preReturn(int sign) {
        return sign > 0 ? 2147483647 : -2147483648;
    }

    bool isAlpha(char c) {
        return isalpha(c) || c == '.';
    }

    int myAtoi(string s) {
        int sign = 1;
        int r = 0;
        bool start = false;
        for (char &c: s) {
            bool isDigit = isdigit(c);

            if ((!start && isAlpha(c)) || (start && !isDigit)) {
                break;
            }

            if (!start) {
                if (c == '+') {
                    sign = 1;
                    start = true;
                    continue;
                } else if (c == '-') {
                    sign = -1;
                    start = true;
                    continue;
                } else if (isDigit) {
                    start = true;
                } else {
                    continue;
                }
            }

            int r0 = c - '0';
            if (preCheck(r, r0, sign)) {
                return preReturn(sign);
            }
            r *= 10;
            r += c - '0';
        }
        return sign * r;
    }
};

File: 008/rust/main.rs
impl Solution {

    pub fn pre_check1(r: u32, r0: u32) -> bool {
         r > 214748364 || (r == 214748364 && r0 > 7)
    }

    pub fn pre_check2(r: u32, r0: u32) -> bool {
         r > 214748364 || (r == 214748364 && r0 >= 8)
    }

    pub fn pre_check(r: u32, r0: u32, sign: i32) -> bool {
        if sign > 0 { Solution::pre_check1(r, r0) } else { Solution::pre_check2(r, r0) }
    }

    pub fn pre_return(sign: i32) -> i32 {
        if sign > 0 { 2147483647 } else { -2147483648 }
    }

    pub fn is_alpha(c: char) -> bool {
        c.is_alphabetic() || c == '.'
    }

    pub fn my_atoi(s: String) -> i32 {
        let mut sign = 1;
        let mut r = 0;
        let mut start = false;
        for c in s.chars() {
            let isDigit = c.is_digit(10);

            if ((!start && Solution::is_alpha(c)) || (start && !isDigit)) {
                break;
            }

            if (!start) {
                if (c == '+') {
                    sign = 1;
                    start = true;
                    continue;
                } else if (c == '-') {
                    sign = -1;
                    start = true;
                    continue;
                } else if (isDigit) {
                    start = true;
                } else {
                    continue;
                }
            }

            let r0 = c.to_digit(10).unwrap();
            if Solution::pre_check(r, r0, sign) {
                return Solution::pre_return(sign);
            }
            r *= 10;
            r += r0;
        }
        return sign * r as i32;
    }
}
